api/src/paths/projects/{projectId}/artifacts/upload.ts

```diff
@@ -0,0 +1,148 @@
+                type: 'object',
+                required: ['file_name', 'encoded_file'],
+                properties: {
+                  file_name: {
+                    type: 'string',
+                    description: 'The name of the file'
+                  },
+                  encoded_file: {
+                    type: 'string',
+                    description: 'The base64 encoded file content'
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  },
+  responses: {
+    200: {
+      description: 'Artifacts upload response.',
+      content: {
+        'application/json': {
+          schema: {
+            type: 'array',
+            items: {
+              type: 'object',
+              properties: {
+                mediaKey: {
+                  type: 'string',
+                  description: 'The S3 unique key for this file.'
+                },
+                lastModified: {
+                  type: 'string',
+                  description: 'The date the object was last modified'
+                }
+              }
+            }
+          }
+        }
+      }
+    },
+    401: {
+      $ref: '#/components/responses/401'
+    },
+    503: {
+      $ref: '#/components/responses/503'
+    },
+    default: {
+      $ref: '#/components/responses/default'
+    }
+  }
+};
+
+/**
+ * Uploads any media in the request to S3, adding their keys to the request, and calling next().
+ *
+ * Does nothing if no media is present in the request.
+ *
+ * TODO: make media handling an extension that can be added to different endpoints/record types
+ *
+ * @returns {RequestHandler}
+ */
+export function uploadMedia(): RequestHandler {
+  return async (req, res, next) => {
+    const debug_label = 'uploadMedia';
+
+    defaultLog.debug({ label: debug_label, message: 'start:', body: req.body });
+
+    if (!req.body.media || !req.body.media.length) {
+      // no media objects included, skipping media upload step
+      return next();
+    }
+
+    const rawMediaArray: IMediaItem[] = req.body.media;
+
+    const s3UploadPromises: Promise<ManagedUpload.SendData>[] = [];
+
+    rawMediaArray.forEach((rawMedia: IMediaItem) => {
+      if (!rawMedia) {
+        return;
+      }
+
+      let media: MediaBase64;
+      try {
+        media = new MediaBase64(rawMedia);
+      } catch (error) {
+        defaultLog.debug({ label: debug_label, message: 'error', error });
+        throw {
+          status: 400,
+          message: 'Included media was invalid/encoded incorrectly'
+        };
+      }
+
+      const metadata = {
+        filename: media.mediaName || '',
+        description: media.mediaDescription || '',
+        date: media.mediaDate || '',
+        username: (req['auth_payload'] && req['auth_payload'].preferred_username) || '',
+        email: (req['auth_payload'] && req['auth_payload'].email) || ''
+      };
+
+      s3UploadPromises.push(uploadFileToS3(media, metadata));
```

Comment:
This needs an `await` otherwise it won't wait and will call `res.status(200)` on line 145 before the upload has actually finished.

We should also check the response, to see if we succeeded or not, and return a different code/response depending on it.

Reply:
That being said, I wouldn't kill yourself trying to find a way to simulate an error or anything.  I would just go based on what the doc says as far as what it will return (as realistically S3 is never itself going to fail and It will be that we send it a bad key, or have a network issue or something).

Reply:
I refactored this file a bit ... I think it's overall in better shape now.